src/maxdiffusion/models/attention_flax.py

```diff
@@ -18,13 +18,237 @@
+    flash_min_seq_length: int = 4096
+    dtype: DType = jnp.float32
+
+    def check_attention_inputs(
+        self,
+        query: Array,
+        key: Array,
+        value: Array
+    ) -> None:
+
+        """Check attention inputs."""
+
+        assert key.ndim == value.ndim, 'k, v must have same rank.'
+        assert query.shape[:-3] == key.shape[:-3] == value.shape[:-3], (
+            'q, k, v batch dims must match.')
+        assert key.shape[-2] == value.shape[-2], ('k, v num_kv_heads must match.')
+        assert key.shape[-3] == value.shape[-3], 'k, v lengths must match.'
+        assert query.shape[-1] == key.shape[-1], 'q, k depths must match.'
+
+    def apply_attention(
+        self,
+        query: Array,
+        key: Array,
+        value: Array,
+        decoder_segment_ids: Array | None = None
+    ):
+        self.check_attention_inputs(query, key, value)
+
+        can_use_flash_attention = query.shape[1] >= self.flash_min_seq_length and key.shape[1] >= self.flash_min_seq_length and value.shape[1] >= self.flash_min_seq_length
+
+        if self.attention_kernel == "dot_product" or self.use_memory_efficient_attention or not can_use_flash_attention:
+            return self.apply_attention_dot(query, key, value)
+        elif self.attention_kernel == "flash":
+            return self.tpu_flash_attention(query, key * self.scale, value, decoder_segment_ids)
+        else:
+            raise ValueError(f'Unexpected attention kernel {self.attention_kernel=}.')
+
+    def tpu_flash_attention(
+        self,
+        query: jax.Array,
+        key: jax.Array,
+        value: jax.Array,
+        decoder_segment_ids: Array | None) -> jax.Array:
+        """TPU Flash Attention"""
+
+        query, kv_size = self.reshape_data_for_flash(query)
+        key, _ = self.reshape_data_for_flash(key)
+        value, _ = self.reshape_data_for_flash(value)
+
+        if decoder_segment_ids is not None:
+            decoder_segment_ids = splash_attention_kernel.SegmentIds(
+            decoder_segment_ids, decoder_segment_ids
+            )
+        axis_names = nn.logical_to_mesh_axes(self.flash_axis_names)
+        segment_axis_names = nn.logical_to_mesh_axes(
+            (BATCH, 'activation_length_no_heads')
+        )
+
+        @functools.partial(
+            shard_map.shard_map,
+            mesh=self.mesh,
+            in_specs=(
+                axis_names,
+                axis_names,
+                axis_names,
+                segment_axis_names
+            ),
+            out_specs=axis_names,
+            check_rep=False
+        )
+        def wrap_flash_attention(query, key, value, decoder_segment_ids):
+            if decoder_segment_ids is not None:
+               assert (
+                    query.shape[2]
+                    == decoder_segment_ids.q.shape[1]
+                ), 'Sharding along sequence dimension not allowed in tpu kernel attention'
+            block_sizes = splash_attention_kernel.BlockSizes(
```

Comment:
Do we want to make block sizes configurable? So, we can experiment on different devices?

Reply:
done